providers/gitlab_test.go

```diff
@@ -39,129 +41,183 @@ func testGitLabBackend() *httptest.Server {
-	b := testGitLabBackend()
-	defer b.Close()
+		DescribeTable("should return expected results",
+			func(in emailsTableInput) {
+				p.AllowUnverifiedEmail = in.allowUnverifiedEmail
+				session := &sessions.SessionState{AccessToken: "gitlab_access_token"}
 
-	bURL, _ := url.Parse(b.URL)
-	p := testGitLabProvider(bURL.Host)
-	p.AllowUnverifiedEmail = true
-	p.Group = "baz"
+				if in.domains != nil {
+					if len(in.domains) >= 1 {
+						p.EmailDomains = in.domains
+					}
+				}
 
-	session := &sessions.SessionState{AccessToken: "gitlab_access_token"}
-	_, err := p.GetEmailAddress(context.Background(), session)
-	assert.NotEqual(t, nil, err)
-}
+				val, err := p.GetEmailAddress(context.Background(), session)
 
-func TestGitLabProviderEmailDomainValid(t *testing.T) {
-	b := testGitLabBackend()
-	defer b.Close()
+				if in.expectedError != nil {
+					Expect(err).To(MatchError(err))
+				} else {
+					Expect(err).To(BeNil())
+					Expect(val).To(Equal(in.expectedValue))
+				}
+			},
+			Entry("unverified email denied", emailsTableInput{
+				expectedError:        errors.New("user email is not verified"),
+				allowUnverifiedEmail: false,
+			}),
+			Entry("unverified email allowed", emailsTableInput{
+				expectedError:        nil,
+				expectedValue:        "foo@bar.com",
+				allowUnverifiedEmail: true,
+			}),
+			Entry("domain valid", emailsTableInput{
+				expectedError:        nil,
+				expectedValue:        "foo@bar.com",
+				allowUnverifiedEmail: true,
+				domains:              []string{"bar.com", "foo.com"},
+			}),
+			Entry("domain invalid", emailsTableInput{
+				expectedError:        errors.New("err"),
+				allowUnverifiedEmail: true,
+				domains:              []string{"baz.com", "foo.com"},
+			}),
+		)
+	})
+
+	Context("when getting username data", func() {
+		It("should return username", func() {
+			session := &sessions.SessionState{AccessToken: "gitlab_access_token"}
+
+			val, err := p.GetUserName(context.Background(), session)
+
+			Expect(err).To(BeNil())
+			Expect(val).To(Equal("FooBar"))
+
+		})
+	})
+
+	Context("when filtering on gitlab entities (groups and projects)", func() {
+		type entitiesTableInput struct {
+			expectedError error
+			expectedValue string
+			projects      []string
+			group         string
+		}
 
-	bURL, _ := url.Parse(b.URL)
-	p := testGitLabProvider(bURL.Host)
-	p.AllowUnverifiedEmail = true
-	p.EmailDomains = []string{"bar.com"}
+		DescribeTable("should return expected results",
+			func(in entitiesTableInput) {
+				p.AllowUnverifiedEmail = true
+				session := &sessions.SessionState{AccessToken: "gitlab_access_token"}
 
-	session := &sessions.SessionState{AccessToken: "gitlab_access_token"}
-	email, err := p.GetEmailAddress(context.Background(), session)
-	assert.Equal(t, nil, err)
-	assert.Equal(t, "foo@bar.com", email)
-}
+				if in.projects != nil {
+					if len(in.projects) >= 1 {
+						p.Projects = in.projects
+					}
+				}
```

Comment:
I think this can be simplified
```suggestion
				if len(in.projects) > 0 {
					p.Projects = in.projects
				}
```

Reply:
Oh yes, that's dumb, I forgot that a nil slice has a length of 0.